File: src/main/scala/impressions.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// scalastyle:off println
//package org.apache.spark.examples.streaming

import scala.sys.process._ // Execute external command

import org.apache.spark.SparkConf
import org.apache.spark.streaming.{Seconds, StreamingContext}
import org.anormcypher._ // spark to neo4j
// without auth

// For TESTing
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.rdd.RDD
import scala.collection.mutable.Queue
//import org.apache.spark.streaming.scheduler.StreamingListenerBatchCompleted;
//import org.apache.spark.streaming.scheduler.*;
//import org.apache.spark.streaming.scheduler.StreamingListener;
import org.apache.log4j.Logger
import org.apache.log4j.Level


/**
 * Counts words in new text files created in the given directory
 * Usage: HdfsWordCount <directory>
 *   <directory> is the directory that Spark Streaming will use to find and read new text files.
 *
 * To run this on your local machine on directory `localdir`, run this example
 *    $ bin/run-example \
 *       org.apache.spark.examples.streaming.HdfsWordCount localdir
 *
 * Then create a text file in `localdir` and the words in the file will get counted.
 */

//private class MyJobListener(ssc: StreamingContext) extends StreamingListener {
//  override def onBatchCompleted(batchCompleted: StreamingListenerBatchCompleted) = synchronized {
//    ssc.stop(true)
//  }
//}

object Impressions {
  def main(args: Array[String]) {
    //if (args.length < 1) {
    //  System.err.println("Usage: HdfsWordCount <directory>")
    //  System.exit(1)
    //}
    val file = "src/main/scala/datadir"
    val outfile = "src/main/scala/out.txt" 

    // suppress INFO
    Logger.getLogger("org").setLevel(Level.WARN)
    Logger.getLogger("akka").setLevel(Level.WARN)

    //StreamingExamples.setStreamingLogLevels()
    val sparkConf = new SparkConf().setAppName("Impressions")
    // Create the context
    val sc = new SparkContext(sparkConf)
    val ssc = new StreamingContext(sc, Seconds(4)) // when using sc
    //val ssc = new StreamingContext(sparkConf, Seconds(2))

    // TEST streaming
    val test_rdd  = sc.textFile("src/main/scala/datadir/test.txt")
    val words = test_rdd.flatMap(_.split(" "))
    val wordCounts = words.map(x => (x, 1)).reduceByKey(_ + _)
    val test_rddQueue: Queue[RDD[(String, Int)]] = Queue()
    test_rddQueue.enqueue( wordCounts)
    val dstream = ssc.queueStream(test_rddQueue)
    dstream.print()

     // Create the FileInputDStream on the directory and use the
     // stream to count words in new files created
    // val lines = ssc.textFileStream(file)
    // val words = lines.flatMap(_.split(" "))
    // val wordCounts = words.map(x => (x, 1)).reduceByKey(_ + _)
    println("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")
    println("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")
    println("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")
    println("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")
    println("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")

    val data1_dns : String = "ec2-52-72-28-165.compute-1.amazonaws.com"
    val neo4j_db : String = "/home/ubuntu/neo4j-community-2.3.2/data/graph.db" 
    val neo4j_db_dir : String = "/home/ubuntu/neo4j-community-2.3.2/data/" 

    // reset database ... the database is in the node data1
    Process("ssh " + data1_dns + " rm -rf " + neo4j_db)!


    // my operator are        schema
    //                        code id    
    // * add new user     |   1        name:string
    // * delete user      |   2        name:string
    // * follow           |   3        follower:string  followed:string
    // * unfollow         |   4        follower:string  followed:string
    // * new tweet        |   5        msg:string  user:string
    // * retweet          |   6        msg:string  user:string original_user:string
   
     


    implicit val connection = Neo4jREST(data1_dns, 7474, neo4j_db_dir)
    val logData = sc.textFile(FILE, 4).cache()
    val count = logData
      .flatMap( _.split(" "))
      .map( w =>
        Cypher("CREATE(:Word {text:{text}})")
          .on( "text" -> w ).execute()
    ).filter( _ ).count()



    ssc.start()
    ssc.awaitTermination()
  }
}
// scalastyle:on println

File: src/impressions.py
#from __future__ import print_function # print rdd
import sys
import time
import json

from pyspark import SparkContext
from pyspark.streaming import StreamingContext
from pyspark.sql import SQLContext # needed to transform PipelinedRDD to RDD for the functions sortByKey
import pyspark

# u = user id
# ts = timestamp
# follows = rdd containing json with the pair (follower, followed)
#def getFollowers(u, ts, follows, unfollows)
#    "get followers from a given user"
#    return 0

def filterNewUser(x):    xj = json.loads(x); return True if ('code' in xj and xj['code'] == 1) else False
def filterDeleteUser(x): xj = json.loads(x); return True if ('code' in xj and xj['code'] == 2) else False
def filterFollow(x):     xj = json.loads(x); return True if ('code' in xj and xj['code'] == 3) else False
def filterUnFollow(x):   xj = json.loads(x); return True if ('code' in xj and xj['code'] == 4) else False
def filterNewTweet(x):   xj = json.loads(x); return True if ('code' in xj and xj['code'] == 5) else False
def filterReTweet(x):    xj = json.loads(x); return True if ('code' in xj and xj['code'] == 6) else False

if __name__ == "__main__":

    # Create a local StreamingContext with two working thread and batch interval of 1 second
    sc = SparkContext("local[2]", "NetworkWordCount")
    ssc = StreamingContext(sc, 1)
    sql = SQLContext(sc)
   
    lines = sc.textFile("./db", 1)
    follows = lines.filter(filterReTweet)
    #lines.filter()
    #counts = lines.map(lambda x: x)
    #              .map(lambda x: (x, 1)) \
    #             .reduceByKey(lambda a, b: a + b) \
    #             .sortByKey(True) 
    # Try this after .reduceByKey: .map(lambda x:(x[0],x[1])) \ # which is faster? I don't know   
    for val in follows.collect(): print val
    #counts.foreach(print)

    #// my operator are        schema
    #//                        code id:short            
    #// * add new user     |   1              id:Long ts:Long name:string
    #// * delete user      |   2              id:Long ts:Long name:string
    #// * follow           |   3              id:Long ts:Long follower_id:Long  followed_id:Long
    #// * unfollow         |   4              id:Long ts:Long follower_id:Long  followed_id:Long
    #// * new tweet        |   5              id:Long ts:Long msg:string  user_id:Long
    #// * retweet          |   6              id:Long ts:Long msg:string  user_id:Long original_user_id:Long
 
    # Create the queue through which RDDs can be pushed to
    # a QueueInputDStream
    #rddQueue = []
    #rddQueue.append(counts)
    
    # Create the QueueInputDStream and use it do some processing
    #inputStream = ssc.queueStream(rddQueue)
    #mappedStream = inputStream.flatMap(lambda x: x.split(' ')) \
    #                          .map(lambda x: (x, 1))
    #reducedStream = mappedStream.reduceByKey(lambda a, b: a + b)
    #sortedStream = reducedStream.sortByKey()
    #inputStream.pprint()
    
    logger = sc._jvm.org.apache.log4j
    logger.LogManager.getLogger("org").setLevel( logger.Level.OFF )
    logger.LogManager.getLogger("akka").setLevel( logger.Level.OFF )
    
    #ssc.start()             # Start the computation
    #ssc.awaitTermination()  # Wait for the computation to terminate
   
    #df = sqlContext.read.json("./db") 

 
    #ssc.start()
    #time.sleep(6)
    #ssc.stop(stopSparkContext=True, stopGraceFully=True)
    sc.stop()


